File: src/main/scala/com/teambr/bookshelf/client/gui/component/control/GuiComponentTextBox.scala
package com.teambr.bookshelf.client.gui.component.control

import com.teambr.bookshelf.client.gui.component.BaseComponent
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.GuiTextField
import org.lwjgl.opengl.GL11

/**
 * This file was created for Bookshelf
 *
 * Bookshelf is licensed under the
 * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
 * http://creativecommons.org/licenses/by-nc-sa/4.0/
 *
 * @author Paul Davis <pauljoda>
 * @since August 04, 2015
 */
abstract class GuiComponentTextBox(x: Int, y: Int, var width: Int, var height: Int)
    extends BaseComponent(x, y) {

    protected var textField = new GuiTextField(0, Minecraft.getMinecraft.fontRendererObj, x, y, width, height)
    /**
     * Called when the value has changed
     * @param value The current text in the field
     */
    def fieldUpdated (value: String) : Unit

    /**
     * Get the current text
     * @return What is in the field
     */
    def getValue : String = textField.getText

    /**
     * Set the value in the field
     * @param str The new string
     */
    def setValue(str: String) : Unit = textField.setText(str)

    /**
     * Get the text field
     * @return The text field (vanilla version)
     */
    def getTextField: GuiTextField = textField

    override def initialize() : Unit = {}

    /**
     * Called when the mouse is pressed
     * @param mouseX Mouse X Position
     * @param y Mouse Y Position
     * @param button Mouse Button
     */
    override def mouseDown(mouseX: Int, y: Int, button: Int) {
        textField.mouseClicked(mouseX, y, button)
        if (button == 1 && textField.isFocused) {
            textField.setText("")
            fieldUpdated(textField.getText)
        }
    }

    /**
     * Used when a key is pressed
     * @param letter The letter
     * @param keyCode The code
     */
    override def keyTyped(letter: Char, keyCode: Int) {
        if (textField.isFocused) {
            textField.textboxKeyTyped(letter, keyCode)
            fieldUpdated(textField.getText)
        }
    }

    override def render(guiLeft: Int, guiTop: Int) {
        GL11.glPushMatrix()
        textField.drawTextBox()
        GL11.glDisable(GL11.GL_ALPHA_TEST)
        GL11.glPopMatrix()
    }

    override def renderOverlay(guiLeft: Int, guiTop: Int) : Unit = {}

    def getWidth: Int = width

    def getHeight: Int = height

    override def isMouseOver(mouseX: Int, mouseY: Int): Boolean = true
}

File: src/main/scala/com/teambr/bookshelf/client/gui/component/control/GuiComponentTextBox.java
package com.teambr.bookshelf.client.gui.component.control;

import com.teambr.bookshelf.client.gui.GuiBase;
import com.teambr.bookshelf.client.gui.component.BaseComponent;
import com.teambr.bookshelf.util.ClientUtils;
import net.minecraft.client.gui.GuiTextField;
import net.minecraft.client.renderer.GlStateManager;
import org.lwjgl.opengl.GL11;

import javax.annotation.Nullable;

/**
 * This file was created for Bookshelf
 * <p>
 * Bookshelf is licensed under the
 * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
 * http://creativecommons.org/licenses/by-nc-sa/4.0/
 *
 * @author Paul Davis - pauljoda
 * @since 2/13/2017
 */
public abstract class GuiComponentTextBox extends BaseComponent {
    // Variables
    protected int width, height;
    protected GuiTextField textField;

    /**
     * Creates the text box
     * @param parent The parent gui
     * @param x The x pos
     * @param y The y pos
     * @param boxWidth The text box width
     * @param boxHeight The text box height, usually 16
     * @param defaultLabel The default label, will translate, can be null
     */
    public GuiComponentTextBox(GuiBase<?> parent, int x, int y, int boxWidth, int boxHeight, @Nullable String defaultLabel) {
        super(parent, x, y);
        this.width = boxWidth;
        this.height = boxHeight;

        textField = new GuiTextField(0, fontRenderer, x, y, width, height);
        if(defaultLabel != null)
            textField.setText(ClientUtils.translate(defaultLabel));
    }

    /*******************************************************************************************************************
     * Abstract Methods                                                                                                *
     *******************************************************************************************************************/

    /**
     * Called when the value in the text box changes
     * @param value The current value
     */
    protected abstract void fieldUpdated(String value);

    /*******************************************************************************************************************
     * BaseComponent                                                                                                   *
     *******************************************************************************************************************/

    /**
     * Called when the mouse is pressed
     * @param x Mouse X Position
     * @param y Mouse Y Position
     * @param button Mouse Button
     */
    @Override
    public void mouseDown(int x, int y, int button) {
        textField.mouseClicked(x, y, button);
        if(button == 1  && textField.isFocused()) {
            textField.setText("");
            fieldUpdated(textField.getText());
        }
    }

    /**
     * Used when a key is pressed
     * @param letter The letter
     * @param keyCode The code
     */
    @Override
    public void keyTyped(char letter, int keyCode) {
        if(textField.isFocused()) {
            textField.textboxKeyTyped(letter, keyCode);
            fieldUpdated(textField.getText());
        }
    }


    /**
     * Called to render the component
     */
    @Override
    public void render(int guiLeft, int guiTop, int mouseX, int mouseY) {
        GlStateManager.pushMatrix();
        textField.drawTextBox();
        GlStateManager.disableAlpha();
        GL11.glDisable(GL11.GL_ALPHA_TEST);
        GlStateManager.popMatrix();
    }

    /**
     * Called after base render, is already translated to guiLeft and guiTop, just move offset
     */
    @Override
    public void renderOverlay(int guiLeft, int guiTop, int mouseX, int mouseY) {
        // NO OP
    }

    /**
     * Used to find how wide this is
     *
     * @return How wide the component is
     */
    @Override
    public int getWidth() {
        return width;
    }

    /**
     * Used to find how tall this is
     *
     * @return How tall the component is
     */
    @Override
    public int getHeight() {
        return height;
    }

    /*******************************************************************************************************************
     * Accessors/Mutators                                                                                              *
     *******************************************************************************************************************/

    public GuiTextField getTextField() {
        return textField;
    }

    public void setTextField(GuiTextField textField) {
        this.textField = textField;
    }
}
